README
NAME
    Geo::Gosmore - Interface to the headless gosmore(1) routing application

SYNOPSIS
    First install gosmore(1), e.g. on Debian:

        sudo aptitude install gosmore

    Then build a gosmore.pak file:

        wget http://download.geofabrik.de/osm/europe/british_isles.osm.bz2
        # pv(1) is not needed, it just shows you the import progress
        bzcat british_isles.osm.bz2 | pv | gosmore rebuild

    Then use this library, with "gosmore_path" being the full path to your
    new gosmore.pak.

        my $gosmore = Geo::Gosmore->new(
            gosmore_path => $gosmore_path,
        );

        my $query = Geo::Gosmore::Query->new(
            flat => '51.5425',
            flon => '-0.111',
            tlat => '51.5614',
            tlon => '-0.0466',
            fast => 1,
            v    => 'motorcar',
        );

        # Returns false if we can't find a route
        my $route = $gosmore->route($query);
        my $distance = $route->distance;

DESCRIPTION
    Provides an interface to the headless version of the gosmore
    <http://wiki.openstreetmap.org/wiki/Gosmore> routing library. When
    compiled with headless support it provides a simple interface to do
    routing. This library just parses its simple output and provides
    accessors for it.

    We also support accessing the headless gosmore(1) program through a
    remote CGI interface.

ATTRIBUTES
  gosmore_method
    Either "binary" or "http". If binary "gosmore_path" is a path to a
    gosmore.pak and we'll invoke gosmore(1) from your $PATH.

    If it's "http" "gosmore_path" is a URL to an online gosmore router.

  gosmore_path
    Either a path to a gosmore.pak (see "gosmore_method") or a HTTP URL to a
    gosmore CGI routing script without query parameters.

METHODS
  route
    Find a route based on the attributes you've passed in. Takes a
    Geo::Gosmore::Query object with your query, returns a
    Geo::Gosmore::Route object.

LICENSE AND COPYRIGHT
    Copyright 2011 Ævar Arnfjörð Bjarmason <avar@cpan.org>

    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Geo::Gosmore

0.05 2011-06-20 15:11:49

    - Support HTTP queries and add documentation

0.04 2011-06-20 08:52:34

    - Add IPC::System::Simple as a dependency

0.03 2011-06-18 19:48:37

    - Documentation, improved API. Travel time API.

0.02 2011-06-18 13:39:17

    - Add the handling of no route found to the API.

0.01 2011-06-18 13:16:41

    - Initial buggy version


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Interface to the headless L<gosmore(1)> routing application',
  'AUTHOR' => 'Ævar Arnfjörð Bjarmason <avar@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'Geo-Gosmore',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Geo::Gosmore',
  'PREREQ_PM' => {
    'Any::Moose' => '0',
    'Cwd' => '0',
    'Data::Dumper' => '0',
    'File::Basename' => '0',
    'Geo::Distance::XS' => '0',
    'IPC::System::Simple' => '0',
    'WWW::Mechanize' => '0',
    'autodie' => '0'
  },
  'VERSION' => '0.05',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Geo/Gosmore/Query.pm
package Geo::Gosmore::Query;
BEGIN {
  $Geo::Gosmore::Query::AUTHORITY = 'cpan:AVAR';
}
BEGIN {
  $Geo::Gosmore::Query::VERSION = '0.05';
}
use Any::Moose;
use warnings FATAL => "all";

has flat => (
    is            => 'ro',
    isa           => 'Num',
    required      => 1,
    documentation => '',
);

has flon => (
    is            => 'ro',
    isa           => 'Num',
    required      => 1,
    documentation => '',
);

has tlat => (
    is            => 'ro',
    isa           => 'Num',
    required      => 1,
    documentation => '',
);

has tlon => (
    is            => 'ro',
    isa           => 'Num',
    required      => 1,
    documentation => '',
);

has fast => (
    is            => 'ro',
    isa           => 'Int',
    required      => 1,
    documentation => '',
);


has v => (
    is            => 'ro',
    isa           => 'Str',
    required      => 1,
    documentation => '',
);

sub query_string {
    my ($self) = @_;

    my @atoms = qw(flat flon tlat tlon fast v);

    my $query_string = join '&', map {
        sprintf "%s=%s", $_, $self->$_;
    } grep {
        defined $self->$_;
    } @atoms;

    return $query_string;
}

1;


lib/Geo/Gosmore/Route.pm
package Geo::Gosmore::Route;
BEGIN {
  $Geo::Gosmore::Route::AUTHORITY = 'cpan:AVAR';
}
BEGIN {
  $Geo::Gosmore::Route::VERSION = '0.05';
}
use Any::Moose;
use warnings FATAL => "all";
use Geo::Distance::XS;

has points => (
    is            => 'ro',
    isa           => 'ArrayRef[ArrayRef[Undef|Num|Str]]',
    required      => 1,
    documentation => '',
);

has distance => (
    is            => 'ro',
    isa           => 'Num',
    documentation => '',
    lazy_build    => 1,
);

sub _build_distance {
    my ($self) = @_;

    my $points = $self->points;
    my $distance = 0;

    my $geo = Geo::Distance->new;

    for (my $i = 1; $i < @$points; $i++) {
        my $prev_point = $points->[$i - 1];
        my $curr_point = $points->[$i];
        my ($lon1, $lat1) = @$prev_point;
        my ($lon2, $lat2) = @$curr_point;

        my $prev_to_curr_distance = $geo->distance(
            kilometer => $lon1, $lat1, $lon2, $lat2,
        );

        $distance += $prev_to_curr_distance;
    }

    return $distance;
}

has travel_time => (
    is            => 'ro',
    isa           => 'Int',
    documentation => '',
    lazy_build    => 1,
);

sub _build_travel_time {
    my ($self) = @_;

    my $first_point = $self->points->[0];
    my $travel_time = $first_point->[4];

    return $travel_time;
}

1;


lib/Geo/Gosmore.pm
package Geo::Gosmore;
BEGIN {
  $Geo::Gosmore::AUTHORITY = 'cpan:AVAR';
}
BEGIN {
  $Geo::Gosmore::VERSION = '0.05';
}
use Any::Moose;
use Any::Moose '::Util::TypeConstraints';
use warnings FATAL => "all";
use autodie qw(:all);
use Geo::Gosmore::Route;
use File::Basename qw(dirname);
use Cwd qw(getcwd);
use WWW::Mechanize;

=encoding utf8

=head1 NAME

Geo::Gosmore - Interface to the headless L<gosmore(1)> routing application

=head1 SYNOPSIS

First install L<gosmore(1)>, e.g. on Debian:

    sudo aptitude install gosmore

Then build a F<gosmore.pak> file:

    wget http://download.geofabrik.de/osm/europe/british_isles.osm.bz2
    # pv(1) is not needed, it just shows you the import progress
    bzcat british_isles.osm.bz2 | pv | gosmore rebuild

Then use this library, with C<gosmore_path> being the full path to
your new F<gosmore.pak>.

    my $gosmore = Geo::Gosmore->new(
        gosmore_path => $gosmore_path,
    );

    my $query = Geo::Gosmore::Query->new(
        flat => '51.5425',
        flon => '-0.111',
        tlat => '51.5614',
        tlon => '-0.0466',
        fast => 1,
        v    => 'motorcar',
    );

    # Returns false if we can't find a route
    my $route = $gosmore->route($query);
    my $distance = $route->distance;

=head1 DESCRIPTION

Provides an interface to the headless version of the
L<gosmore|http://wiki.openstreetmap.org/wiki/Gosmore> routing
library. When compiled with headless support it provides a simple
interface to do routing. This library just parses its simple output
and provides accessors for it.

We also support accessing the headless L<gosmore(1)> program through a
remote CGI interface.

=head1 ATTRIBUTES

=cut

enum GosmoreMethod => qw(
    binary
    http
);

=head2 gosmore_method

Either C<binary> or C<http>. If binary L</gosmore_path> is a path to a
F<gosmore.pak> and we'll invoke L<gosmore(1)> from your C<$PATH>.

If it's C<http> L</gosmore_path> is a URL to an online gosmore router.

=cut

has gosmore_method => (
    is            => 'ro',
    isa           => 'GosmoreMethod',
    required      => 1,
    default       => "binary",
    documentation => "The gosmore method to use. Either 'binary' or 'http'",
);

=head2 gosmore_path

Either a path to a F<gosmore.pak> (see L</gosmore_method>) or a HTTP
URL to a gosmore CGI routing script without query parameters.

=cut

has gosmore_path => (
    is            => 'ro',
    isa           => 'Str',
    required      => 1,
    documentation => "The full path to the gosmore.pak file to use, or a HTTP URL with a gosmore instance we can send queries to",
);

has _gosmore_dirname => (
    is            => 'ro',
    isa           => 'Str',
    documentation => "The full path to the directory the gosmore.pak file is in",
    lazy_build    => 1,
);

sub _build__gosmore_dirname {
    my ($self) = @_;

    my $gosmore_path    = $self->gosmore_path;
    my $gosmore_dirname = dirname($gosmore_path);

    return $gosmore_dirname;
}

has _mech => (
    is            => 'ro',
    isa           => 'WWW::Mechanize',
    documentation => "Our instance of WWW::Mechanize",
    lazy_build    => 1,
);

sub _build__mech {
    my ($self) = @_;

    WWW::Mechanize->new(
        user_agent => __PACKAGE__,
    );
}

=head1 METHODS

=cut

=head2 route

Find a route based on the L<attributes|/ATTRIBUTES> you've passed
in. Takes a L<Geo::Gosmore::Query> object with your query, returns a
L<Geo::Gosmore::Route> object.

=cut

sub route {
    my ($self, $query) = @_;

    my $lines = $self->_get_normalized_routing_lines($query);

    my @points;
    for my $line (@$lines) {
        # We couldn't find a route
        return if $line eq 'No route found';

        print STDERR "$line\n" if $ENV{DEBUG};

        # We're getting a stream of lat/lon values
        next unless $line =~ /^[0-9]/;

        my ($lat, $lon, $junction_type, $style, $remaining_time, $name) = split /,/, $line;
        push @points => [ $lat, $lon, $junction_type, $style, $remaining_time, $name ];
    }

    my $route = Geo::Gosmore::Route->new(
        points => \@points,
    );

    return $route;
}

sub _get_normalized_routing_lines {
    my ($self, $query) = @_;

    my $method = $self->gosmore_method;
    my @lines;

    my $query_string = $query->query_string;
    if ($method eq 'binary') {
        my $gosmore_dirname = $self->_gosmore_dirname;

        local $ENV{QUERY_STRING} = $query_string;
        local $ENV{LC_NUMERIC} = "en_US";
        my $current_dirname = getcwd();
        chdir $gosmore_dirname;
        open my $gosmore, "gosmore |";
        chdir $current_dirname;

        while (my $line = <$gosmore>) {
            # Skip the HTTP header
            next if $. == 1 || $. == 2;

            $line =~ s/[[:cntrl:]]//g;

            push @lines => $line;
        }

    } elsif ($method eq 'http') {
        my $mech = $self->_mech;
        my $url = sprintf "%s?%s", $self->gosmore_path, $query_string;
        $mech->get($url);
        my $content = $mech->content;
        use Data::Dumper;
        @lines = map {
            s/[[:cntrl:]]//g;
            $_;
        } split /\n/, $content;
    }

    return \@lines;
}

1;

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Ævar Arnfjörð Bjarmason <avar@cpan.org>

This program is free software, you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;

